design/anonymousTokens.md

```diff
@@ -0,0 +1,96 @@
+# Auth2 Anonymous tokens
+
+## Problem statement
+
+Currently KBase users must log into KBase in order to view a narrative or download (DL) data from
+a narrative. We wish for users that are not logged in to be able to do both.
+
+## Background
+
+* When a Narrative docker container is spun up for a user, the narrative launcher expects a KBase
+  auth token and assigns the container to the user.
+* JSON workspace object DLs go through the Data Import Export (DIE) server, which appears
+  to handle anonymous requests correctly based on examination of the code (would need to be
+  tested).
+* Other DLs go through the Narrative Job Server (NJS), which requires a token - and all KBase
+  Software Development Kit (SDK) apps require a token - to create the DL package in Shock.
+  * DIE is responsible for DLing the package from Shock for the user in the case of
+    non-staging area DLs
+      * DIE appears to handle anonymous DLs correctly based on examination of the
+        code (would need testing).
+      * HOWEVER, the Shock node would need to be publicly readable, and no DL apps
+        make that change (because they all expect a token).
+
+## Document purpose
+
+Propose a preliminary design for KBase auth service support of anonymous tokens (ATs)
+that allow anonymous users to have a transient identity when interacting with KBase. Discuss how
+those tokens would interact with the rest of the KBase infrastructure and what changes would be
+necessary.
+
+## Anonymous tokens
+
+* The KBase auth service would be enhanced to offer AT support.
+* ATs would have the same fields as regular tokens except that the the `type` would
+  always be `Anonymous`.
+* ATs would have a random user name assigned upon creation.
+  * Could be a concatenated dictionary style name for mnemonic purposes
+    (e.g. [https://gfycat.com/shoddydismalamoeba](https://gfycat.com/shoddydismalamoeba)).
+      * Need to check collision frequency.
+* ATs would exist in a separate token and user namespace from standard tokens.
+* ATs could not be used to create other tokens.
```

Comment:
Good idea, but might cause trickiness if we're using them to start download jobs.

Reply:
Yeah, I considered this, and I think a combination of 1) having a fairly long lifetime for the token and 2) the FE getting a new token / userid if the token is within X hours of expiration should take care of if. Download jobs shouldn't take that long, usually.

The alternative is making short lived anonymous agent tokens, but that adds scope.